sherpa_contrib/diag_resp.py

```diff
@@ -0,0 +1,908 @@
+#
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301 USA.
+#
+
+
+
+
+""" Create diagonal RMF and flat ARF objects of value unity.
+
+Helper routines for creating a diagonal RMF and flat ARF response
+objects for various telescopes and instruments.
+The default is to generate Chandra/ACIS responses for PI-type spectra
+extracted from ACIS stowed background event files.
+
+Examples
+--------
+
+>>> from sherpa_contrib.diag_resp import *
+>>> diagrmf,flatarf = mkdiagresp()
+>>>
+>>> set_rmf(diagrmf, bkg_id=1)
+>>> set_arf(flatarf, bkg_id=1)
+
+or
+
+>>> diagrmf,flatarf = mkdiagresp(refspec="bkg.pi")
+
+or
+
+>>> bkgspec = get_bkg()
+>>> diagrmf,flatarf = mkdiagresp(refspec=bkgspec)
+
+"""
+
+__revision__ = "22 July 2024"
+
+import os
+import warnings
+from logging import getLogger
+from functools import wraps
+from re import sub
+
+import numpy.typing as npt
+from numpy import arange
+
+from ciao_contrib._tools.fileio import get_keys_from_file
+
+from sherpa.astro.data import DataPHA
+from sherpa.astro.ui import create_rmf as shpmkrmf
+from sherpa.astro.ui import create_arf as shpmkarf
+
+
+
+def _reformat_wmsg(func):
+    """
+    Decorator:
+
+    Trim extraneous user-warning information showing path and line number
+    where Sherpa warning message is being thrown for replacing 0s with ethresh.
+    """
+
+    @wraps(func)
+    def run_func(*args, **kwargs):
+        warnings_reset = warnings.formatwarning
+        warnings.formatwarning = lambda msg, *args_warn, **kwargs_warn: f'Warning: {msg}\n'
+
+        func_out = func(*args,**kwargs)
+
+        warnings.formatwarning = warnings_reset
+
+        return func_out
+
+    return run_func
+
+
+
+def _quash_shpverb(func):
+    """
+    Decorator:
+
+    Temporarily quash Sherpa verbosity.
+    """
+
+    def run_func(*args, **kwargs):
+        sherpalog = getLogger("sherpa")
+        loglevel = sherpalog.level
+        sherpalog.setLevel(0)
+
+        func_out = func(*args,**kwargs)
+
+        sherpalog.setLevel(loglevel)
+
+        return func_out
+
+    return run_func
+
+
+
+def _get_random_string(strlen: int = 16) -> str:
+    from string import ascii_letters,digits
+    from random import choices
+
+    chars = ascii_letters + digits
+
+    return ''.join(choices(chars, k=strlen))
+
+
+
+def _arg_case(arg: str|int|None=None,
```

Comment:
I don't think this will work with python 3.9, since the `|` form was, I believe, added in Python 3.10.

So change to `...(arg: Optional[Union[str, int]],` and `...) -> Optional[str]` and add a `from typing import Optional, Union` (assuming they are not already imported).

Reply:
There are other lines below which will need updating too.

Reply:
having installed a 3.9 environment, you are unfortunately correct, and I have revised... it looked so nice using pipes